### PR TITLE
add serverless apis

### LINKED ISSUE #68: Supprt deploy serverless inference endpoint

### PR DESCRIPTION
**Pull request by SeanHH86**
fix #68  and Reorg apis router

### CODE REVIEW

File: common/types/model.go [hunk 1]
```diff
@@ -169,7 +169,18 @@ type ModelStatusEventData struct {
 }
 
 const (
-	SpaceType     = 0
-	InferenceType = 1
-	FinetuneType  = 2
+	SpaceType      = 0 // space
```
Comment (Rader) [inline common/types/model.go:172]: SpaceType = iota

File: api/handler/model.go [hunk 2]
```diff
@@ -1132,3 +1189,196 @@ func (h *ModelHandler) AllFiles(ctx *gin.Context) {
...
+		req.Revision = "main" // default repo branch
+	}
+
+	if req.MinReplica < 0 || req.MaxReplica < 0 || req.MinReplica > req.MaxReplica {
+		slog.Error("Bad request setting for replica", slog.Any("MinReplica", req.MinReplica), slog.Any("MaxReplica", req.MaxReplica))
+		httpbase.BadRequest(ctx, "Bad request setting for replica")
+		return
+	}
+
+	deployReq := types.DeployActReq{
```
Comment (ganisback) [inline api/handler/model.go:1250]: security level should be force to public

File: component/repo.go [hunk 3]
```diff
@@ -1955,10 +1995,43 @@ func (c *RepoComponent) checkDeployPermission(ctx context.Context, repoType type
 	return &user, repo, deploy, nil
 }
 
-func (c *RepoComponent) DeployUpdate(ctx context.Context, repoType types.RepositoryType, namespace, name, currentUser string, deployID int64, req types.ModelRunReq) error {
-	_, _, deploy, err := c.checkDeployPermission(ctx, repoType, namespace, name, currentUser, deployID)
+func (c *RepoComponent) checkDeployPermissionForServerless(ctx context.Context, deployReq types.DeployActReq) (*database.User, *database.Repository, *database.Deploy, error) {
```
Comment (Rader) [inline component/repo.go:1998]: only `admin` has permission to operate on serverless deployment

File: api/handler/model.go [hunk 4]
```diff
@@ -1132,3 +1189,196 @@ func (h *ModelHandler) AllFiles(ctx *gin.Context) {
...
+	}
+
+	err = h.c.DeployStop(ctx, stopReq)
+	if err != nil {
+		slog.Error("Failed to stop deploy", slog.Any("error", err))
+		httpbase.ServerError(ctx, err)
+		return
+	}
+	httpbase.OK(ctx, nil)
+}
```
Comment (ganisback) [inline api/handler/model.go:1384]: need add delete api for serverless

File: api/handler/model.go (hunk 2 above)
Comment (SeanHH86) [reply api/handler/model.go:1250]: fixed.

File: builder/deploy/deployer.go [hunk 5]
```diff
@@ -154,7 +169,7 @@ func (d *deployer) dedicatedDeploy(ctx context.Context, dr types.DeployRepo) (*d
 func (d *deployer) buildDeploy(ctx context.Context, dr types.DeployRepo) (*database.Deploy, error) {
 	var deploy *database.Deploy = nil
 	var err error = nil
-	if dr.SpaceID > 0 {
+	if dr.SpaceID > 0 || dr.Type == types.ServerlessType {
```
[the remark this answers is not in the document]
Comment (SeanHH86) [reply builder/deploy/deployer.go]: fixed.

File: common/types/model.go (hunk 1 above)
Comment (SeanHH86) [reply common/types/model.go]: fixed.

File: component/repo.go (hunk 3 above)
Comment (SeanHH86) [reply component/repo.go:1998]: add function to check admin

**Review by ganisback — APPROVED** [review]
LGTM

File: common/types/model.go [hunk 6]
```diff
@@ -169,10 +169,10 @@ type ModelStatusEventData struct {
 }
 
 const (
-	SpaceType      = 0 // space
-	InferenceType  = 1 // inference endpoint
-	FinetuneType   = 2 // finetune
-	ServerlessType = 3 // serverless
+	SpaceType      = iota // space
+	InferenceType  = 1    // inference endpoint
+	FinetuneType   = 2    // finetune
```
Comment (Rader) [inline common/types/model.go:174]: don't need to set values manually if `iota` is used

File: builder/deploy/deployer.go [hunk 7]
```diff
@@ -85,7 +85,7 @@ func newDeployer(s scheduler.Scheduler, ib imagebuilder.Builder, ir imagerunner.
 
 func (d *deployer) GenerateUniqueSvcName(dr types.DeployRepo) string {
 	uniqueSvcName := ""
-	if dr.SpaceID > 0 {
+	if dr.Type == types.InferenceType {
```
Comment (Rader) [inline builder/deploy/deployer.go:88]: should be `SpaceType` here

**Review by Rader — CHANGES_REQUESTED** [review]

**Review by Rader — APPROVED** [review]

**Pull request merged**